doc/design/clustered-etcd-alert-store/design.md

```diff
@@ -0,0 +1,150 @@
+# Clustered Etcd Alert Store
+
+Type: Design Document (and PR)
+
+Date: 2019-09-20
+
+Author: David C Wang <dcwangmit01@gmail.com>
+
+## Status Quo
+
+The current Alertmanager clustering implementation will allow multiple
+instances of Alertmanager to sync recently-updated notification log entries and
+alert silences via the gossip protocol.  The syncing of alert state is not yet
+implemented.
+
+Currently, Alertmanager alert state is entirely stored in process memory.  It
+is lost upon restart of Alertmanager.  This, in conjunction with the fact that
+Alertmanager alert state is not synced with clustered peers presents several
+problems:
+
+* Alert emitters must send alerts to every single Alertmanager instance in the
+  cluster.  If some Alertmanager instances receive alerts that other instances
+  do not, their state will drift apart.
+
+  * Typically, the alert emitter is Prometheus which uses service discovery to
+    locate all Alertmanager instances, and then sends alerts to each individual
+    target.  Custom alert emitters may not have that service discovery
+    capability and must be explicity configured.
```

Comment:
Is this under the assumption that the *alert emitter* is not under the control of the administrator and thus can not be easily changed to emit alerts not to one but to all Alertmanagers?

Reply:
Yes.  That that is a very real deployment model.  If a vendor does not support metrics and only supports network notifications, we may ask them to format and send their events as *alerts* that are compatible with the Alertmanager API. 

Since these vendors would be outside of our Kubernetes cluster and would not have access to Alertmanagers in our private kubernetes overlay network, the only way they can send *alerts* to our Alermanagers is to traverse via Kubernetes Loadbalancer integration, which would hit exactly one target.

My team has already written a few alert-exporter type software components that impedance match between fault-based (not metrics-based) legacy systems and Alertmanager.  Some of these tools can read custom API faults tables and translate them directly into alerts sent to Alertmanager.  We even convert SNMP Traps to Alertmanager alerts.  Painful.

Reply:
That sounds like a case for a custom alert forwarder, which doesn't need to live in the AM itself.